src/OpenTelemetry.Exporter.Geneva/MsgPackExporter/MsgPackTraceExporter.cs

```diff
@@ -243,6 +243,13 @@ internal int SerializeActivity(Activity activity)
             cntFields += 1;
         }
 
+        if (!string.IsNullOrEmpty(activity.TraceStateString))
```

Comment:
could we make this opt-in? `Activity` does some special mechanisms (walk the parent chain...), to find tracestatestring, so it may not be free! (That was one reason why it was removed in OTel LogRecord).

I'd suggest to do benchmarks and see if this affects perf. (make sure to test with A1->A2->A3 scenario with tracestate set on A1 only/similar)

Reply:
It's already made into Part B. I'm not sure if it's right to make it opt-in. If anything I feel like it's should be more of an opt-out if a customer is concerned of the performance impact.

I'll run the benchmarks and test its effects on performance.

Can you link me to more details regarding the removal in OTel LogRecord?

Reply:
> Can you link me to more details regarding the removal in OTel LogRecord?

https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Logs/ILogger/OpenTelemetryLoggerOptions.cs#L78

Reply:
Thanks, Cijo. I have found the original discussion on it: https://github.com/open-telemetry/opentelemetry-dotnet/pull/4334#discussion_r1160873655

But the difference is that the use of tracestate on log record is "probably rare" and more importantly,  [TraceState is not on log data model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#log-and-event-record-definition).

However, traceState is in Span of Part B.

Reply:
I did some benchmark tests (see code in PR). Here is also some data with those tests for with/without the code change in src files.

# Without code change in MsgPackTraceExporter
| Method                                       | Mean     | Error   | StdDev  | Gen0   | Allocated |
|--------------------------------------------- |---------:|--------:|--------:|-------:|----------:|
| ExportActivity                               | 276.2 ns | 2.20 ns | 2.06 ns |      - |         - |
| SerializeActivity                            | 240.1 ns | 4.41 ns | 6.74 ns |      - |         - |
| SerializeActivityWithoutTraceState           | 253.9 ns | 2.16 ns | 2.02 ns |      - |         - |
| SerializeActivityWithTraceState              | 257.7 ns | 2.30 ns | 2.15 ns |      - |         - |
| SerializeActivityWithTraceStateInGrandparent | 258.0 ns | 3.10 ns | 2.59 ns |      - |         - |
| CreateActivityWithGenevaExporter             | 441.7 ns | 8.79 ns | 7.79 ns | 0.0396 |     416 B |


# With code change in MsgPackTraceExporter
| Method                                       | Mean     | Error   | StdDev  | Gen0   | Allocated |
|--------------------------------------------- |---------:|--------:|--------:|-------:|----------:|
| ExportActivity                               | 266.2 ns | 0.34 ns | 0.28 ns |      - |         - |
| SerializeActivity                            | 243.1 ns | 1.42 ns | 1.26 ns |      - |         - |
| SerializeActivityWithoutTraceState           | 248.7 ns | 0.78 ns | 0.65 ns |      - |         - |
| SerializeActivityWithTraceState              | 273.8 ns | 3.75 ns | 3.32 ns |      - |         - |
| SerializeActivityWithTraceStateInGrandparent | 280.2 ns | 5.54 ns | 5.69 ns |      - |         - |
| CreateActivityWithGenevaExporter             | 452.1 ns | 4.50 ns | 3.99 ns | 0.0396 |     416 B |




# Without code change in TldTraceExporter
| Method                                           | Mean       | Error     | StdDev    | Allocated |
|------------------------------------------------- |-----------:|----------:|----------:|----------:|
| MsgPack_SerializeActivity                        | 239.387 ns | 0.6162 ns | 0.4811 ns |         - |
| TLD_SerializeActivity                            | 283.528 ns | 1.9925 ns | 1.8638 ns |         - |
| TLD_SerializeActivityWithoutTraceState           | 293.012 ns | 3.6684 ns | 3.2519 ns |         - |
| TLD_SerializeActivityWithTraceState              | 293.466 ns | 1.1386 ns | 1.0093 ns |         - |
| TLD_SerializeActivityWithTraceStateInGrandparent | 285.946 ns | 0.9807 ns | 0.9174 ns |         - |
| MsgPack_ExportActivity                           | 267.778 ns | 1.0824 ns | 1.0125 ns |         - |
| TLD_ExportActivity                               |   1.050 ns | 0.0124 ns | 0.0110 ns |         - |

# With code change in TldTraceExporter
| Method                                           | Mean       | Error     | StdDev    | Allocated |
|------------------------------------------------- |-----------:|----------:|----------:|----------:|
| MsgPack_SerializeActivity                        | 239.467 ns | 1.1202 ns | 0.9930 ns |         - |
| TLD_SerializeActivity                            | 279.810 ns | 1.2504 ns | 0.9763 ns |         - |
| TLD_SerializeActivityWithoutTraceState           | 294.145 ns | 1.2952 ns | 1.1481 ns |         - |
| TLD_SerializeActivityWithTraceState              | 319.608 ns | 5.9573 ns | 5.5724 ns |         - |
| TLD_SerializeActivityWithTraceStateInGrandparent | 314.195 ns | 1.1738 ns | 0.9164 ns |         - |
| MsgPack_ExportActivity                           | 265.354 ns | 1.0098 ns | 0.9445 ns |         - |
| TLD_ExportActivity                               |   1.102 ns | 0.0469 ns | 0.0482 ns |         - |

The weird thing is that if tracestate is in grandparent activity, sometimes it's faster than if it's in the grandchild activity.

Reply:
> could we make this opt-in?

+1

Reply:
285.946 ns to 314.195 ns is ~10% perf hit. (Most likely, the deeper the chain is, the more the perfhit is)

> But the difference is that the use of tracestate on log record is "probably rare"
I am not sure if that is the case or not.. It could be!

> However, traceState is in Span of Part B.
That does not prevent us from making it opt-in.

Reply:
> That does not prevent us from making it opt-in.

I see. I'll add an option.

Meanwhile, just for the sake of discussion, I agree with your claim that deeper chain could cause more perfhit because the implementation of `Activity.TraceStateString`. However, there are two strange things contrary to my understanding:
1. in many cases, `TLD_SerializeActivityWithTraceStateInGrandparent` is faster than `TLD_SerializeActivityWithTraceState` in the same run. 285.946 ns < 293.466 ns is an example and I've got other examples after running `TraceExporterBenchmarks` multiple times.
2. I step debugged the unit test to `var activity = source.StartActivity("SayHello", ActivityKind.Internal, parentActivity.Context, null, links)` line, and found that right after this line, VS shows that `activity.Parent` is `null` and `TraceStateString` got `parentActivity`'s value `"some=state"`. Instead, I was expecting at least `activity.Parent` points to `parentActivity`. (`activity.TraceStateString` might have got its value while it's being displayed.)

I'm not sure what I'm missing.

Reply:
I'm more coming from security/privacy side - if users bump the exporter version and start to see tracestate leaking EUII and there is no easy way to turn it off, they might be very upset.